src/Entity/EntityIdValue.php

```diff
@@ -138,27 +156,45 @@ public function getArrayValue() {
 	 */
 	public static function newFromArray( $data ) {
 		if ( !is_array( $data ) ) {
-			throw new IllegalValueException( '$data must be an array; got ' . gettype( $data ) );
+			throw new IllegalValueException( '$data must be an array' );
 		}
 
-		if ( !array_key_exists( 'entity-type', $data ) ) {
-			throw new IllegalValueException( "'entity-type' field required" );
+		if ( array_key_exists( 'id', $data ) ) {
+			return self::newFromIdSerialization( $data['id'] );
+		} elseif ( array_key_exists( 'entity-type', $data ) && array_key_exists( 'numeric-id', $data ) ) {
+			return self::newFromNumericId( $data['entity-type'], $data['numeric-id'] );
+		} else {
+			throw new IllegalValueException( "Either 'id' or 'entity-type' and 'numeric-id' fields required" );
 		}
+	}
 
-		if ( !array_key_exists( 'numeric-id', $data ) ) {
-			throw new IllegalValueException( "'numeric-id' field required" );
+	/**
+	 * @param string $idSerialization
+	 *
+	 * @throws IllegalValueException
+	 * @return self
+	 */
+	private static function newFromIdSerialization( $idSerialization ) {
+		try {
```

Comment:
I did not wanted to waste time just for a nicer message. I wanted this do die hard.